homeassistant/components/shelly/switch.py

```diff
@@ -0,0 +1,165 @@
+"""Representation of Z-Wave switches."""
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.const import CONF_DEVICE_ID
+from homeassistant.core import callback
+
+from .const import (
+    COMMAND_SUFFIX,
+    CONF_MODEL,
+    CONF_TOPIC,
+    DOMAIN,
+    MODEL_SWITCHES,
+    MODELS,
+)
+
+ON = "on"
+OFF = "off"
+TOGGLE = "toggle"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Shelly switch(es) from config entry."""
+
+    device_id = config_entry.data[CONF_DEVICE_ID]
+    topic = config_entry.data[CONF_TOPIC]
+    model = config_entry.data[CONF_MODEL]
+    switches_count = MODELS[model][MODEL_SWITCHES]
+    switches = []
+    for i in range(0, switches_count):
+        switch_topic = topic + f"relay/{i}"
+        if i == 0:
+            unique_id = device_id
+            via_device = None
+        else:
+            unique_id = f"{device_id}_{i+1}"
+            via_device = device_id
+        switches.append(
+            ShellySwitch(
+                hass.components.mqtt, unique_id, switch_topic, model, via_device
+            )
+        )
+
+    async_add_entities(switches)
+
+
+class ShellySwitch(SwitchEntity):
+    """Representation of a Shelly switch."""
+
+    def __init__(self, mqtt, unique_id, topic, model, via_device):
+        """Initialize the MQTT switch."""
+        self._mqtt = mqtt
+        self._unique_id = unique_id
+        self._via_device = via_device
+        self._topic = topic
+        self._model = model
+        self._power = None
+        self._energy = None
+        self._unsubscribe = None
+        self._state = None
+
+    async def async_added_to_hass(self):
+        """Subscribe to MQTT events."""
+        await super().async_added_to_hass()
+        await self._subscribe_topics()
+
+    async def _subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+
+        @callback
+        def state_message_received(msg):
+            """Handle new MQTT state messages."""
+            if msg.topic == self._topic:
+                payload = msg.payload
+                if payload == ON:
+                    self._state = True
+                elif payload == OFF:
+                    self._state = False
+                # how do we handle the "overpower" state?
```

Comment:
I think that `overpower` should be a separate entity of domain `binary_sensor`.

Reply:
Interesting thought. It can also be an attribute, but the main question what to do with on/off in this case? Leave the previous value?

Reply:
As far as I know, the attributes should be static values. So `power`, `energy`, `overpower`, `overtemperature` shouldn't be attributes only separate entities.
State `overpower` is only when relay is `on` so state of the `switch` entity shouldn't change in that situation.

Reply:
Power and energy are mentioned as properties in the [developer docs for the switch platform](https://developers.home-assistant.io/docs/core/entity/switch#properties).
I know that generally there's a push towards independent sensors, but I'm not sure it applies for these. Other examples are the light entities with the light level, Climate entities with temp, etc...